target/linux/mpc85xx/image/Makefile

```diff
@@ -1,107 +1,92 @@
 		cd $(KDIR_TMP)/empty-initramfs; \
 		find . | cpio -o -H newc | gzip -9n > $(KDIR_TMP)/empty-initramfs.gpio.gz
 
 	-$(STAGING_DIR_HOST)/bin/mkimage -A $(LINUX_KARCH) -O linux -T ramdisk \
-		-C gzip -n "empty initramfs" \
+		-C gzip -n "$(DEVICE_PROFILE) empty initramfs" \
 		-d $(KDIR_TMP)/empty-initramfs.gpio.gz \
-		$(KDIR_TMP)/$(IMG_PREFIX)-empty-initramfs.image
-endef
+		$(KDIR_TMP)/$(IMG_PREFIX)-empty-initramfs.img
 
-define Image/BuildKernel
-	cp $(KDIR)/zImage $(BIN_DIR)/$(IMG_PREFIX)-zImage
-	$(foreach dts,$(DTS_TARGETS),
-		$(LINUX_DIR)/scripts/dtc/dtc -I dts -O dtb $(DTS_DIR)/$(dts).dts > $(BIN_DIR)/$(IMG_PREFIX)-$(notdir $(dts)).fdt
-	)
+	-dd if=$(KDIR_TMP)/$(IMG_PREFIX)-empty-initramfs.img >> $@
 endef
 
-define Image/BuildKernel/Initramfs
-	cp $(KDIR)/zImage-initramfs $(BIN_DIR)/$(IMG_PREFIX)-initramfs-zImage
-ifeq ($(SUBTARGET),generic)
-	cp $(KDIR)/cuImage.tl-wdr4900-v1-initramfs $(BIN_DIR)/$(IMG_PREFIX)-tl-wdr4900-v1-initramfs.uImage
-endif
-ifeq ($(SUBTARGET),p1020)
-	cp $(KDIR)/zImage-initramfs $(BIN_DIR)/$(IMG_PREFIX)-hiveap-330-initramfs.zImage
-endif
+define Build/copy-file
+	cat "$(1)" >> "$@"
 endef
 
-define Image/Build/Aerohive
-	$(call Build/empty-initramfs)
-ifeq ($(2),hiveap-330)
-	( \
-		dd if=$(BIN_DIR)/$(IMG_PREFIX)-$(2).fdt bs=256k conv=sync; \
-		dd if=$(KDIR_TMP)/$(IMG_PREFIX)-empty-initramfs.image bs=256k conv=sync; \
-		dd if=$(BIN_DIR)/$(IMG_PREFIX)-root.$(1) bs=40704k conv=sync; \
-		dd if=$(KDIR)/zImage conv=sync; \
-	) > $(BIN_DIR)/$(IMG_PREFIX)-$(2)-sysupgrade.img
-endif
+define Build/dtb
+	$(call Image/BuildDTB,$(DTS_DIR)/$(DEVICE_DTS).dts,$@.dtb,,--space $(DTB_SIZE))
+	cp $@.dtb $(BIN_DIR)/$(IMG_PREFIX)-$(DEVICE_NAME).fdt
 endef
 
-define Image/Build/TPLINK
-	-$(STAGING_DIR_HOST)/bin/mktplinkfw \
-		-H $(4) -W $(5) -F $(6) -N OpenWrt -V $(REVISION) $(7) \
-		-k $(KDIR)/$(3) \
-		-r $(KDIR)/root.$(1) \
-		-o $(call factoryname,$(1),$(2))
-	-$(STAGING_DIR_HOST)/bin/mktplinkfw \
-		-H $(4) -W $(5) -F $(6) -N OpenWrt -V $(REVISION) $(7) -s \
-		-k $(KDIR)/$(3) \
-		-r $(KDIR)/root.$(1) \
-		-o $(call sysupname,$(1),$(2))
+define Build/tplink-v1-image
+	$(STAGING_DIR_HOST)/bin/mktplinkfw \
+		-H $(TPLINK_HWID) -W $(TPLINK_HWREV) -F $(TPLINK_LAYOUTID) \
+		-N "$(VERSION_DIST)" -V $(REVISION) $(1) \
+		-k $@ \
+		-r $(KDIR)/root.squashfs \
+		-o $@.new
+	@mv $@.new $@
 endef
 
-define Image/Build/Profile/hiveap-330
-	$(call Image/Build/Aerohive,$(1),hiveap-330)
-endef
+ifeq ($(SUBTARGET),generic)
 
-define Image/Build/Profile/TLWDR4900
-	$(call Image/Build/TPLINK,$(1),tl-wdr4900-v1,cuImage.tl-wdr4900-v1,0x49000001,1,16Mppc)
+define Device/TLWDR4900
+  DEVICE_TITLE := TP-Link TL-WDR4900
+  DEVICE_PACKAGES := kmod-ath9k wpad-mini
+  DEVICE_PROFILE := tl-wdr4900-v1
+  DEVICE_DTS := $$(DEVICE_PROFILE)
+  BLOCKSIZE := 128k
+  TPLINK_HWID := 0x49000001
+  TPLINK_HWREV := 1
+  TPLINK_LAYOUTID := 16Mppc
+  KERNEL := dtb | copy-file $(LINUX_DIR)/arch/$(LINUX_KARCH)/boot/cuImage.$$(DEVICE_PROFILE)
```

Comment:
I originally was trying to get KERNEL_NAME working, but as cuImage is within the boot directory, it is never exported. The way the old style dealt with this was by using https://github.com/lede-project/source/blob/master/target/linux/mpc85xx/Makefile#L27 and https://github.com/lede-project/source/blob/master/include/kernel-defaults.mk#L126. If anyone has a good way to deal with this besides the method above, please let me know.